webapp/pages/admin/users.vue

```diff
@@ -1,15 +1,173 @@
+        <ds-flex-item width="50px">
+          <ds-button @click="next" :disabled="!hasNext" icon="arrow-right" primary />
+        </ds-flex-item>
+        <ds-flex-item width="50px">
+          <ds-button @click="back" :disabled="!hasPrevious" icon="arrow-left" primary />
+        </ds-flex-item>
+      </ds-flex>
+    </ds-card>
+    <ds-card v-else>
+      <ds-placeholder>
+        {{ $t('admin.users.empty') }}
+      </ds-placeholder>
+    </ds-card>
+  </div>
 </template>
 
 <script>
-import HcEmpty from '~/components/Empty.vue'
+import gql from 'graphql-tag'
+import isemail from 'isemail'
 
 export default {
-  components: {
-    HcEmpty,
+  data() {
+    const pageSize = 15
+    return {
+      offset: 0,
+      pageSize,
+      first: pageSize,
+      User: [],
+      hasNext: false,
+      email: null,
+      filter: null,
+      form: {
+        formData: {
+          query: '',
+        },
+      },
+    }
+  },
+  computed: {
+    hasPrevious() {
+      return this.offset > 0
+    },
+    fields() {
+      return {
+        index: '#',
+        name: this.$t('admin.users.table.columns.name'),
+        slug: this.$t('admin.users.table.columns.slug'),
+        createdAt: this.$t('admin.users.table.columns.createdAt'),
+        contributionsCount: {
+          label: '🖉',
+          align: 'right',
+        },
+        commentedCount: {
+          label: '🗨',
+          align: 'right',
+        },
+        shoutedCount: {
+          label: '❤',
+          align: 'right',
+        },
+        role: {
+          label: this.$t('admin.users.table.columns.role'),
+          align: 'right',
+        },
+      }
+    },
+  },
+  apollo: {
+    User: {
+      query() {
+        return gql(`
+        query($filter: _UserFilter, $first: Int, $offset: Int, $email: String) {
+          User(email: $email, filter: $filter, first: $first, offset: $offset, orderBy: createdAt_desc) {
+            id
+            name
+            slug
+            role
+            createdAt
+            contributionsCount
+            commentedCount
+            shoutedCount
+          }
+        }
+      `)
+      },
+      variables() {
+        const { offset, first, email, filter } = this
+        const variables = { first, offset }
+        if (email) variables.email = email
+        if (filter) variables.filter = filter
+        return variables
+      },
+      update({ User }) {
+        if (!User) return []
+        this.hasNext = User.length >= this.pageSize
+        if (User.length <= 0 && this.offset > 0) return this.User // edge case, avoid a blank page
+        return User.map((u, i) => Object.assign({}, u, { index: this.offset + i }))
+      },
+    },
+  },
+  methods: {
+    back() {
+      this.offset = Math.max(this.offset - this.pageSize, 0)
+    },
+    next() {
+      this.offset += this.pageSize
+    },
+    submit(formData) {
+      this.offset = 0
+      const { query } = formData
+      if (isemail.validate(query)) {
```

Comment:
I guess you explored the option of using our styleguides' async validators? https://github.com/yiminghe/async-validator
and maybe couldn't because it doesn't have to be an email, it can be other input as well?

`isemail` is an interesting package, it's used by more than 450k people, but only 189 stars :woman_shrugging:

Reply:
Yeah, I found `isemail` in our `yarn.lock when I `git grep 'email'`. You're right: I want to check if it's an email and I don't want to throw an error or sth. in case it's not an email.

Reply:
great! thanks for walking us through the process! That is a great way to decide when to add a package or not :smile_cat: